### ISSUE TITLE
MPC coordination between parts needs improvement

### ISSUE THREAD

**Issue by adriatic** [monterey-framework/prototyper#22]
This issue is a collection of several problems:

#### 1. 
Click on autocomplete clear or, deletion of the last character typed in should clear the content of the `version grid` **and** disable the `Add this pair` button.

#### 2. 
The button `Remove selected packages` should be enabled only when the collector grid is non-empty.

#### 3.
The loop deleting selected packages in the collector component crashes invoking the `removeSelectedPackages` method, as the statement `if ((item as any).fSelected) ` is syntactically incorrect (it is cast to `any` and that type does not have `fSelected` attribute.

#### 4. 
The newly added buttons

![image](https://cloud.githubusercontent.com/assets/2712405/21581548/90b7204a-d003-11e6-9653-f465d879e72e.png)

are created for testing, I presume (particularly the outcome using `semver` modifier (this is a great idea, @Thanood )
.

**Comment by Thanood** [comment]
Regarding 1: I agree and I think that *any* interaction with `autocomplete` should clear the version list.

Regarding 2: Agreed.

Regarding 3: That's not entirely true, as the cast to `any` is valid in this case. `item` is of type `Object` which doesn't have the `fSelected` attribute. `any` on the other hand, has "defined" any attributes you could come up with. The error states that `item` itself in undefined, not `fSelected`.

Did you encounter this **without** duplicate packages? So far I've only seen this with duplicate packages. For example when I adding five jQueries and and trying to remove the last four.

Anyway, I think I'm not using KendoUI's `datasource` correctly.

Regarding 4: Yes, these are created for testing the api of `mpc-component` (which atm consists of a single data-bound property `packages`). It's for debugging the communication with the "outside world".
The semver operator applies when getting the selected packages, its default being "^" as  think this is used in most cases.

There are a few more issues:
- adding an existing package twice (which souldn't be possible anyway) sometimes leads to this item being added another time (leading to four additions of the same package)
- clearing the `autocomplete` leads to an error in `PackageService` (the `response` is not `response.ok`)
- when adding a lot of packages the collector grid grows larger than its container

![image](https://cloud.githubusercontent.com/assets/677826/21581628/0b1c0270-d03a-11e6-968c-18c3cf0ed984.png)


- `add predefined packages` always concats these packages with the currently set array of packages. Clicking it twice leads to six packages being added

**Comment by adriatic** [comment]
Comments to @Thanood comments:

> clearing the autocomplete leads to an error in PackageService (the response is not response.ok)

I believe that this is the item 1 in my [original post](https://github.com/monterey-framework/prototyper/issues/22#issue-198268912)

> when adding a lot of packages the collector grid grows larger than its container

This is the consequence of not setting the `pagesize` limit - similar to what I [set for the version chooser](https://github.com/monterey-framework/prototyper/blob/master/app/src/mpc/basic-components/version-chooser.ts#L21)

**Comment by Thanood** [comment]
> This is the consequence of not setting the pagesize limit

Thanks. Added `pagesize` limit now.
I'm currently looking at improving the "synchronization" of `packages` and the KendoUI datasource which should solve most of the issues regarding adding/removing items.
Also I've found the `datasource.getByUid()` function which makes everything a bit easier. 😃 

Turns out that you cannot simply combine `@bindable()` and `BindingEngine.collectionObserver`, at least when the `@bindable()` array is modified from another `@bindable()` in its parent.

At the moment there's a `@bindable()` chain (two-way databinding):
`screen.packages` => `mpc-component.packages` => `collector-component.packages`

When you change `screen.packages` f.i. by using `push`, the property `collector-component.packages` is not notified of the change, even if you observe the collection directly with Aurelia's `BindingEngine`.

**Comment by Thanood** [comment]
I've seen your changes to your original post just now.
Sorry, I kept the page open as GitHub adds new messages to the thread without a reload but it doesn't update posts when someone edits them. 😃

**Comment by Thanood** [comment]
@adriatic please see #23

**Comment by adriatic** [comment]
> When you change `screen.packages` f.i. by using `push`, the property `collector-component.packages` is not notified of the change, even if you observe the collection directly with Aurelia's `BindingEngine`.

Is this not a bug? Let's see @JeroenVinke's opinion before creating this issue.

**Comment by Thanood** [comment]
Just reproduced it in this gist: https://gist.run?id=5e93cd78f2b9569aff54d96efb077bf3
And @JeroenVinke  was fater. 😏

**Comment by adriatic** [comment]
@JeroenVinke - would you please help me by writing a high level description of a fix like you just did - as looking at the diff is not always the best way to understand what took place 😄

**Comment by JeroenVinke** [comment]
Basically, the `collectionObserver` of the BindingEngine watches for changes in the `packages` array. So whenever this array is modified (using array functions like .push() and pop()) the `packagesChanged` function is executed. When the MPC adds something to the `packages` array ([here](https://github.com/monterey-framework/prototyper/blob/ccb54d00c76e86532cccb31dfa3b6b8ea2d30f06/app/src/mpc/screen.ts#L14)) the collection observer calls `packagesChanged`.

But like @Thanood mentioned in Gitter, the collectionobserver is lost when the instance of the array changes, so the collectionobserver may not be the cleanest solution for the problem

**Comment by adriatic** [comment]
Thank you. Based on your explanation, once could say that Daniel made "illegal" use of the collectionobserver, as it does not support the case where the **instance** of the array changes (likely because in that case its reference to the observed array becomes bad). The whole construct in this situation looks very fragile, as the collectionobserver should not ever "get lost".

**Comment by Thanood** [comment]
I would agree on "illegal" since any `@bindable()` can (infact even should) be modified from the outside in a way the array can be overwritten instead of modified.

**Comment by adriatic** [comment]
One more "synchronization improvement issue"

![image](https://cloud.githubusercontent.com/assets/2712405/21583356/363a6aee-d04b-11e6-83ac-b9dafaba9693.png)

___unless we might ever consider importing multiple versions of the same package - something I cannot see the reason for.___

**Comment by Thanood** [comment]
Interesting.. I've tested it and it worked for me.
Can you check if the autocomplete triggers the `filtering` event? Maybe it even throws an error which doesn't happen here (for whatever reason).

**Comment by adriatic** [comment]
I believe that my problem statement was a bit tricky: it is true that the click on autocomplete's 'x' resets the name and version choosers  - but the click on 'add this pair' button **does not**. I claim tha it should !

**Comment by Thanood** [comment]
Now I see.. 😃 
Thanks for clarifying.

**Comment by Thanood** [comment]
Hmm after clearing autocomplete...

![image](https://cloud.githubusercontent.com/assets/677826/21587707/7ddcee00-d0df-11e6-907a-862c85ab6a92.png)


"add this pair" is disabled.
Strange.
